README.md
# CSRFImmunity

## Description

CSRFImmunity is a NODE JS module to prevent Cross Site Request Forgery issue in web applications. This module can be used to generate and validate CSRF tokens. It can also be used to generate token based on various user defined parameters like page name, form action, userid or username and any other parameters the application needs. The module also uses a JSON string to generate the CSRF token which is the most important part of this module, it makes sure that the token is random enough for an user and a page. The application developer needs to make sure that the parameter in the JSON argument is unique for each user and each page.

## How to use

The module requires one encryption key to encrypt the CSRF token. This key need be passed to the module while initiating it in your code. Please make sure that the key is random not predictable and big in length.

```
var csrfi = require(csrfimmunity).setKeyForEncryption("encryption_key");
```

The setKeyForEncryption function makes sure that different application uses different and personal secret keys.

## Methods Available

### generateToken(options)

The above method accepts a JSON object as an argument. The JSON object is then used to generate the CSRF tokens. You can pass anything you need in this JSON object to make the CSRF token specific to certain page, form and user. An example of the options parameter is :

```
var options = '{"pagename":"login","formaction":"https:\/\/www.weekendsecurity.org", "username":"piyushpattanayak"}';

csrfi.generateToken(options, function(csrftoken){
    // Work with the token here
});
```

Please use a minimum of two parameter in the JSON argument which should be unique as per the user and per page. In the above example the "pagename" parameter is unique to the web page and the "username" parameter is unique to current logged in user.

### generateTimeBasedToken(options)

This method also accepts the exact same JSON parameter, but generates a time based CSRF token, which means this token will also have the timestamp when it is created. You can use this method to generate time based CSRF tokens and then use the varifyTimeBasedToken to validate this CSRF token.

```
var options = '{"pagename":"login","formaction":"https:\/\/www.weekendsecurity.org", "username":"piyushpattanayak"}';

csrfi.generateTimeBasedToken(options, function(csrftoken){
    // Work with the token here
});
```

The CSRF Token will somewhat look like :

*30466274c8c59179efffe210d72f9802aafa73a7f4462eed2781b9f80d42c9937d9687320f1cc5e7e1d1112a757bc0013a97166d663bb9ded09f0379e257988b77b6f32bc3c089a3a4b77c3050ec0d09fb74e77714bc468eea34dcadac1a4e4776c719d27c8416aa160e6d3bb6fb1af6dcd4b38b7c608cdce3e44b3158a2*

This token can directly be used by as a hidden parameter in the HTML form parameter.

### verifyToken(csrfToken, options)

This method is used to verify the CSRF token from the page. It accepts two parameters, the csrf token and the JSON object with the same key and value pair used during the generateToken function. Sending a different JSON string as argument will fail the check and will return false.

#### Example

```
var options = '{"pagename":"login","formaction":"https:\/\/www.weekendsecurity.org", "username":"piyushpattanayak"}';

var status = csrfi.verifyToken(csrfToken, options); (This will return boolean);
```

### verifyTimeBasedToken(formToken, cookieToken, options)

This method is used to verify time based CSRF tokens. You need to have add an extra key value pair in the JSON object as *"milliseconds"* and assign its value in milliseconds. This will validate the token age based on the time provided. If you do not pass the milliseconds key, then the function will choose the default time which is 250000 milliseconds.

#### Example

```
var options = '{"pagename":"login","formaction":"https:\/\/www.weekendsecurity.org", "username":"piyushpattanayak"}';

csrfi.generateTimeBasedToken(options, function(csrftoken){
    // Get the CSRF Token
});

var optionsForVerification = var options = '{"pagename":"login","formaction":"https:\/\/www.weekendsecurity.org", "username":"piyushpattanayak","milliseconds":"300000"}';

var status = csrfi.verifyTimeBasedToken(csrfToken, optionsForVerification); (This will return boolean)
```

## An overall example for this module

```
var csrf = require('../csrfimmunity').setKeyForEncryption("cdshcjkbdsckdslkclkncd");

var options = '{"pagename":"login", "username":"piyushpattanayak"}';

csrf.generateTimeBasedToken(options, function(timeBasedToken){
    console.log(timeBasedToken); // This will print the time based token
});

csrf.generateToken(options, function(plainToken){
    console.log(plainToken); // This will print usual token
});


var optionsNew = '{"pagename":"login", "username":"piyushpattanayak", "milliseconds":"3000000"}';

var statusTimeBased = csrf.verifyTimeBasedToken(timeBasedToken, optionsNew);
console.log(statusTimeBased); // This will print the boolean success or failure

var statusPlainToken = csrf.verifyToken(plainToken, options);
console.log(statusPlainToken); // This will print the boolean success or failure
```


## Author

- **Name:** Piyush Pattanayak
- **Email :** pattanayak.piyush@gmail.com


CHANGELOG.md
## CHANGE LOG

### 0.2.1

Bug Fixes and code improvements

### 0.2.0

Support time based CSRF token


test/test.js
var Assert = require('assert');
var csrfimmunity = require('../index').setKeyForEncryption("password123");
var options = {};

describe(__filename, function() {

    before(function(next) {
        // process.chdir(__dirname + '/test/app');
        options.formname = 'loginform';
        options.actionurl = 'http://www.weekendsecurity.org/login';
        options = JSON.stringify(options);
        next();
    });

    it('should match with the CSRF token with timestamp', function(next){
        csrfimmunity.generateTimeBasedToken(options, function(token){
            var result = csrfimmunity.verifyTimeBasedToken(token, options);
            Assert.equal(result, true);
            next();
        });
    });

    it('should match with the CSRF token without timestamp', function(next){
        csrfimmunity.generateToken(options, function(token){
            var result = csrfimmunity.verifyToken(token, options);
            Assert.equal(result, true);
            next();
        });
    });

    it('should fail the CSRF token check after time expires', function(done){
        var newoptions = JSON.parse(options);
        newoptions.milliseconds = 1;

            csrfimmunity.generateTimeBasedToken(options, function(token){
                setTimeout(function(){
                    var result = csrfimmunity.verifyTimeBasedToken(token, JSON.stringify(newoptions));
                    Assert.equal(result, false);
                    done();
                }, 3000);
            });
    });
});


test/failcase.js
var Assert = require('assert');
var csrfimmunity = require('../index').setKeyForEncryption("password123");
var options = {};

describe(__filename, function() {

    before(function(next) {
        options.formname = 'loginform';
        options.actionurl = 'http://www.weekendsecurity.org/login';
        options = JSON.stringify(options);
        next();
    });

    it('should match with the CSRF token with timestamp without options', function(next) {
        csrfimmunity.generateTimeBasedToken(null, function(token) {
            var result = csrfimmunity.verifyTimeBasedToken(token, null);
            Assert.equal(result, true);
            next();
        });
    });

    it('should match with the CSRF token without timestamp and options', function(next) {
        csrfimmunity.generateToken(null, function(token) {
            var result = csrfimmunity.verifyToken(token, null);
            Assert.equal(result, true);
            next();
        });
    });

    it('Blank csrftoken should return false for non-timestamp', function(next) {
        csrfimmunity.generateToken(null, function(token) {
            var result = csrfimmunity.verifyToken('', null);
            Assert.equal(result, false);
            next();
        });
    });

    it('Blank csrftoken should return false for timestamp', function(next) {
        csrfimmunity.generateTimeBasedToken(null, function(token) {
            var result = csrfimmunity.verifyTimeBasedToken('', null);
            Assert.equal(result, false);
            next();
        });
    });

    it('Should return false when options is not a valid JSON for time based token', function(next) {
        try {
            csrfimmunity.generateTimeBasedToken('test', function(token) {
                var result = csrfimmunity.verifyTimeBasedToken('test', null);
                next();
            });
        } catch (e) {
            Assert.equal(e.toString(), 'Error: Options is not a JSON String');
            next();
        }
    });

    it('Should return false when options is not a valid JSON for non time based token', function(next) {
        try {
            csrfimmunity.generateToken('test', function(token) {
                var result = csrfimmunity.verifyToken('test', null);
                next();
            });
        } catch (e) {
            Assert.equal(e.toString(), 'Error: Options is not a JSON String');
            next();
        }
    });

    it('Should return null if encryption key is blannk or null', function(next) {
        var csrf = require('../index').setKeyForEncryption('');
        Assert.equal(csrf, null);
        next();
    });

    it('Should return false when options is not a valid JSON for non time based token verify function', function(next) {

        csrfimmunity.generateToken(options, function(token) {
            try {
                var result = csrfimmunity.verifyToken(token, 'testing');
                next();
            } catch (e) {
                Assert.equal(e.toString(), 'Error: Options is not a JSON String');
                next();
            }
        });
    });

    it('Should return false when options is not a valid JSON for time based token verify function', function(next) {

        csrfimmunity.generateTimeBasedToken(options, function(token) {
            try {
                var result = csrfimmunity.verifyTimeBasedToken(token, 'testing');
                next();
            } catch (e) {
                Assert.equal(e.toString(), 'Error: Options is not a JSON String');
                next();
            }
        });
    });
});


csrfImmunity.js
var randomstring = require("randomstring"),
    encryptDecrypt = require("./lib/encrypt-decrypt"),
    separator = "###",
    defaultTimeDiff = 250000;


function isJsonString(str) {
    try {
        var js = JSON.parse(str);
        if(Object.keys(js).length < 2){
          console.warn("CSRFImmunity Module (csrfImmunity.js) : Options parameter is not enough. Application might be risky to CSRF attacks");
        }
    } catch (e) {
      console.log(e);
      console.log("CSRFImmunity Module (csrfImmunity.js) : Not a valid JSON string, " + str);
      throw new Error(e);
    }
    return true;
}

module.exports = {

    setKeyForEncryption: function(key){
      if(key !== "" && key !== null){
        encryptDecrypt.setKeyForEncryption(key);
        return this;
      }else{
        console.log("Key not found. Please provide the encryption key");
        return null;
      }
    },

    generateTimeBasedToken: function(options){

      //options needed to generate the csrf token Ex : {"formname":"loginform", formaction="http://www.weekendsecurity.org/submit"}
      options = options || {"null":"null"};

      //Check for valid JSON string
      isJsonString(options);

      //Time calculated to check the CSRF token validity
      var milliseconds = (new Date).getTime();

      //A random String to pass in the CSRF token and cookie value
      var randomString = randomstring.generate(15);

      //Converting the options JSON to hexadecimal value
      var optionsBuffer = new Buffer(options);
      var encodedOptions = optionsBuffer.toString('hex');

      //Collecting all the token parameters for token generations
      var tokenHidden = encodedOptions + separator + randomString + separator + milliseconds;

      // Encrypting the CSRF token
      var token = encryptDecrypt.encrypt(tokenHidden);

      return token;
    },

    generateToken: function(options){
      options = options || {"null":"null"};

      //Check for valid JSON string
      isJsonString(options);

      //Time calculated to check the CSRF token validity
      var milliseconds = (new Date).getTime();

      //A random String to pass in the CSRF token and cookie value
      var randomString = randomstring.generate(15);

      //Converting the options JSON to hexadecimal value
      var optionsBuffer = new Buffer(options);
      var encodedOptions = optionsBuffer.toString('hex');

      //Collecting all the token parameters for token generations
      var tokenHidden = encodedOptions + separator + randomString;

      // Encrypting the CSRF token
      var token = encryptDecrypt.encrypt(tokenHidden);

      return token;
    },

    verifyTimeBasedToken: function(csrfToken, options){

      //The options which is used to generate the CSRF token
      options = options || '{"null":"null"}';

      //Check for valid JSON string
      isJsonString(options);

      //Need the CSRF token to be passed here for verification
      csrfToken = csrfToken || "";

      //Return false if the csrfToken and cookieValue are sent null
      if(csrfToken === ""){
        return false;
      }

      //Decrypting the csrf token
      try{
        var decTokenHidden = encryptDecrypt.decrypt(csrfToken);
      }catch (e) {
          console.log(e);
          console.log("CSRFImmunity (csrfImmunity.js) : CSRF decryption failed.");
          throw new Error(e);
      }

      //Split the tokens using the separators and store it in an array
      var decTokenHiddenArray = decTokenHidden.split(separator);

      //Converting the options in the hidden token to ascii and then into Javascript object
      var optionsBuffer = new Buffer(decTokenHiddenArray[0], 'hex');
      var optionsjsFromToken = JSON.parse(optionsBuffer.toString('ascii'));
      var optionsjsFromParam = JSON.parse(options);

      //Fetching the time limit of token validity from options parameter.
      var timeDef = optionsjsFromParam.milliseconds || defaultTimeDiff;

      //Verifying if the CSRF token with time, and options parameter
      try{
          var currentMilliseconds = (new Date).getTime();

          if(currentMilliseconds < Number(decTokenHiddenArray[2])+ timeDef){
              for (key in optionsjsFromToken) {
                  if(optionsjsFromToken[key] !== optionsjsFromParam[key]){
                    return false;
                  }
              }
              return true;
          }else{
            return false;
          }
      }catch (e){
        console.log(e);
        throw new Error(e);
      }

    },

    verifyToken: function(csrfToken, options){

          //The options which is used to generate the CSRF token
          options = options || '{"null":"null"}';

          isJsonString(options);

          //Need the CSRF token to be passed here for verification
          csrfToken = csrfToken || "";

          //Return false if the csrfToken and cookieValue are sent null
          if(csrfToken === ""){
            return false;
          }

          //Decrypting the csrf token
          try{
            var decTokenHidden = encryptDecrypt.decrypt(csrfToken);
          }catch (e) {
              console.log(e);
              console.log("CSRFImmunity (csrfImmunity.js) : CSRF decryption failed.");
              throw new Error(e);
          }

          //Split the tokens using the separators and store it in an array
          var decTokenHiddenArray = decTokenHidden.split(separator);

          //Converting the options in the hidden token to ascii and then into Javascript object
          var optionsBuffer = new Buffer(decTokenHiddenArray[0], 'hex');
          var optionsjsFromToken = JSON.parse(optionsBuffer.toString('ascii'));
          var optionsjsFromParam = JSON.parse(options);

          //Verifying if the CSRF token with time, and options parameter
          try{

            for (key in optionsjsFromToken) {
                if(optionsjsFromToken[key] !== optionsjsFromParam[key]){
                  return false;
                }
            }
          }catch (e){
            console.log(e);
            throw new Error(e);
          }
          return true;
    }

}


lib/encrypt-decrypt.js
var crypto = require('crypto'),
    algorithm = 'aes-256-ctr',
    encryptionKey = '';

var colors = require('colors/safe');

module.exports = {

    setKeyForEncryption: function(keyForm){
      encryptionKey = keyForm;
    },

    encrypt: function(text){
      password = encryptionKey;

      if(password === ''){
        console.log(colors.red("Key not provided for encryption"));
        return null;
      }
      var cipher = crypto.createCipher(algorithm,password);
      var crypted = cipher.update(text,'utf8','hex');
      crypted += cipher.final('hex');
      return crypted;
    },

    decrypt: function(text){
      password = encryptionKey;

      if(password === ''){
        console.log(colors.red("Key not provided for decryption"));
        return null;
      }
      var decipher = crypto.createDecipher(algorithm,password);
      var dec = decipher.update(text,'hex','utf8');
      dec += decipher.final('utf8');
      return dec;
    }

};


index.js
var colors = require('colors/safe');
var randomstring = require("randomstring"),
    encryptDecrypt = require("./lib/encrypt-decrypt"),
    separator = "###",
    defaultTimeDiff = 250000;

function isJsonString(str) {
    try {
        var js = JSON.parse(str);
        if (Object.keys(js).length < 2) {
            console.warn(colors.red("CSRFImmunity Module (csrfImmunity.js) : Options parameter is not enough. Application might be risky to CSRF attacks"));
        }
    } catch (e) {
        console.log(e);
        console.log(colors.red("CSRFImmunity Module (csrfImmunity.js) : Not a valid JSON string, " + str));
        // throw new Error(e);
        return false;
    }
    return true;
}

module.exports = {

    setKeyForEncryption: function(key) {
        if (key !== "" && key !== null) {
            encryptDecrypt.setKeyForEncryption(key);
            return this;
        } else {
            console.log(colors.red("Key not found. Please provide the encryption key"));
            return null;
        }
    },

    generateTimeBasedToken: function(options, callback) {

        //options needed to generate the csrf token Ex : {"formname":"loginform", formaction="http://www.weekendsecurity.org/submit"}
        options = options || '{"null": "null"}';

        //Check for valid JSON string
        if(isJsonString(options)){
            //Time calculated to check the CSRF token validity
            var milliseconds = new Date().getTime();

            //A random String to pass in the CSRF token and cookie value
            var randomString = randomstring.generate(15);

            //Converting the options JSON to hexadecimal value
            var optionsBuffer = new Buffer(options);
            var encodedOptions = optionsBuffer.toString('hex');

            //Collecting all the token parameters for token generations
            var tokenHidden = encodedOptions + separator + randomString + separator + milliseconds;

            // Encrypting the CSRF token
            var token = encryptDecrypt.encrypt(tokenHidden);

            callback(token);
        }else{
            throw new Error('Options is not a JSON String');
        }
    },

    generateToken: function(options, callback) {
        options = options || '{"null": "null"}';

        //Check for valid JSON string
        if(isJsonString(options)){
            //Time calculated to check the CSRF token validity
            var milliseconds = new Date().getTime();

            //A random String to pass in the CSRF token and cookie value
            var randomString = randomstring.generate(15);

            //Converting the options JSON to hexadecimal value
            var optionsBuffer = new Buffer(options);
            var encodedOptions = optionsBuffer.toString('hex');

            //Collecting all the token parameters for token generations
            var tokenHidden = encodedOptions + separator + randomString;

            // Encrypting the CSRF token
            var token = encryptDecrypt.encrypt(tokenHidden);

            callback(token);
        }else{
            throw new Error('Options is not a JSON String');
        }
    },

    verifyTimeBasedToken: function(csrfToken, options) {

        //The options which is used to generate the CSRF token
        options = options || '{"null":"null"}';

        //Check for valid JSON string
        if(isJsonString(options)){
            //Need the CSRF token to be passed here for verification
            csrfToken = csrfToken || "";

            //Return false if the csrfToken and cookieValue are sent null
            if (csrfToken === "") {
                return false;
            }

            //Decrypting the csrf token
            try {
                var decTokenHidden = encryptDecrypt.decrypt(csrfToken);

                //Split the tokens using the separators and store it in an array
                var decTokenHiddenArray = decTokenHidden.split(separator);

                //Converting the options in the hidden token to ascii and then into Javascript object
                var optionsBuffer = new Buffer(decTokenHiddenArray[0], 'hex');
                var optionsjsFromToken = JSON.parse(optionsBuffer.toString('ascii'));
                var optionsjsFromParam = JSON.parse(options);

                //Fetching the time limit of token validity from options parameter.
                var timeDef = optionsjsFromParam.milliseconds || defaultTimeDiff;

                var currentMilliseconds = new Date().getTime();

                //Verifying if the CSRF token with time, and options parameter
                if (currentMilliseconds < Number(decTokenHiddenArray[2]) + timeDef) {
                    for (var key in optionsjsFromToken) {
                        if (optionsjsFromToken[key] !== optionsjsFromParam[key]) {
                            return false;
                        }
                    }
                    return true;
                } else {
                    return false;
                }
            } catch (e) {
                console.log(colors.red(e));
                console.log(colors.red("CSRFImmunity (index.js) : CSRF decryption failed.", e));
                throw new Error(e);
            }
        }else{
            throw new Error('Options is not a JSON String');
        }
    },

    verifyToken: function(csrfToken, options) {

        //The options which is used to generate the CSRF token
        options = options || '{"null":"null"}';

        if(isJsonString(options)){
            //Need the CSRF token to be passed here for verification
            csrfToken = csrfToken || "";

            //Return false if the csrfToken and cookieValue are sent null
            if (csrfToken === "") {
                return false;
            }

            //Decrypting the csrf token
            try {
                var decTokenHidden = encryptDecrypt.decrypt(csrfToken);
                //Split the tokens using the separators and store it in an array
                var decTokenHiddenArray = decTokenHidden.split(separator);

                //Converting the options in the hidden token to ascii and then into Javascript object
                var optionsBuffer = new Buffer(decTokenHiddenArray[0], 'hex');
                var optionsjsFromToken = JSON.parse(optionsBuffer.toString('ascii'));
                var optionsjsFromParam = JSON.parse(options);

                //Verifying if the CSRF token with time, and options parameter
                for (var key in optionsjsFromToken) {
                    if (optionsjsFromToken[key] !== optionsjsFromParam[key]) {

                        return false;
                    }
                }

                return true;
            } catch (e) {
                console.log(e);
                console.log(colors.red("CSRFImmunity (csrfImmunity.js) : CSRF decryption failed."));
                throw new Error(e);
            }
        }else{
            throw new Error('Options is not a JSON String');
        }
    }

};
